File: quint/prolog/lists/p23.pl
% P23 (**) Extract a given number of randomly selected elements from a list.

:- ensure_loaded(p20).

rnd_select(_,0,[]).
rnd_select(X,N,[Y|Ys]) :-
    N > 0,
    length(X, Xl),
    M is random(Xl) + 1,
    % Remove a random element and add it to the result.
    remove_at(Y,X,M,Zs),
    Nn is N - 1,
    rnd_select(Zs,Nn,Ys).

File: ben/haskell/Q21-28-lists-again/p23.hs
-- Extract a given number of randomly selected elements from a list.

import System.Random

rnd_select :: [a] -> Int -> IO [a]
rnd_select list i = do g <- getStdGen
                       let nbs = take i (randomRs (0, length list - 1) g)
                           in return (foldr (\a b -> (list!!a):b) [] nbs) 

-- Somewhat nicer
rnd_select' :: [a] -> Int -> IO [a]
rnd_select' list i = do g <- getStdGen
                        return $ take i [list!!n | n <- (randomRs (0, length list - 1) g)]
